### PR TITLE
[SDL-0274] Add preferred FPS to VideoStreamingCapability

### LINKED ISSUE #1553: [SDL 0274] Add preferred FPS to VideoStreamingCapability
Proposal: [Add preferred FPS to VideoStreamingCapability](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0274-add-preferred-FPS.md)

> This proposal extends SDL video streaming feature by adding preferred FPS to `VideoStreamingCapability`, so that frame rate of video streaming can be adjusted per OEM.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/913

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with revisions.  The revisions will include the items listed in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/913#issuecomment-579062117) from the author.

> The proposal .md file was updated to reflect these revisions on 2/10/2020.

### PR DESCRIPTION
**Pull request by shiniwat**
Fixes #1553, #1860 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Ran SDLVideoStreamingCapabilitySpec

#### Core Tests
- Core 4.5 (w/o having preferredFPS parameter)
- Core 6.1 (w/ preferredFPS parameter)

Core version / branch / commit hash / module tested against: Commit to Core 6.1 is pending
HMI name / version / branch / commit hash / module tested against: Our own HMI w/ Video streaming capability enabled.

### Summary
This PR adds preferredFPS parameter to VideoStreamingCapability, and modified SDLStreamingVideoLifecycleManager accordingly. 

##### Breaking Changes
* none

##### Enhancements
* SDL-0274

##### Bug Fixes
* issue #1553

### Tasks Remaining:
- [sdl_core issue #3243]

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by joeljfischer** [comment]
Hi @shiniwat, this proposal has not been earmarked for a release, so this PR has cannot be merged until that release. At the earliest that would be October. I can keep this PR open if you wish, but be aware that it will not be reviewed at this time.

**Comment by shiniwat** [comment]
Understood. Thanks for letting it me know.

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m
Comment (yoooriii) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.m:399]: One should not compare a bool value with YES or NO, especially with YES:
1. The method -isEqualToString: already produces a bool value (YES or NO) which can be used directly without comparing it one more time to YES.
2. Let's take a more sophisticated example:
const BOOL bValue = (BOOL)2; // true condition
if (bValue) { /* true condition works */ }
if (bValue == YES) { /* true condition does not work */ }

Comment (shiniwat) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Thanks for the review. Fixed as suggested.

**Comment by shiniwat** [comment]
Because SDL 0323 implementation needs to handle preferredFPS and max bitrate in the same way, and preferredFPS requires this PR, I have added the SDL 0323 implementation in this PR at [this commit](https://github.com/smartdevicelink/sdl_ios/pull/1613/commits/a2a77b21ef90412ccfebc55c0371f8d3e7d5153f).

File: SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m
Comment (joeljfischer) [inline SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:413]:
```suggestion
```

**Comment by joeljfischer** [comment]
@shiniwat Please link to the SDL Core and HMI PRs that were used to test this PR.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (joeljfischer) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.m:33]: This is incorrect, the initializer should be deprecated and replaced. Please generate the file using the RPC generator and the correct branch of the rpc_spec. You'll need to modify the existing file instead of replacing it with the new file, but it will ensure that the new pieces are created correctly.

**Review by joeljfischer — CHANGES_REQUESTED** [review]
I cannot review the video lifecycle manager piece until the Core PR is approved, but I have a few notes I saw when reviewing other pieces.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.m]: Because RPC generator did not exist when I did initially pull-requested (which was in March), SDLVideoStreamingCapability was manually edited as you pointed out. I have updated the code based on RPC generator's generated code. Thanks.

**Comment by shiniwat** [comment]
> @shiniwat Please link to the SDL Core and HMI PRs that were used to test this PR.

Hi @joeljfischer, here are the links:
sdl_core PR : https://github.com/smartdevicelink/sdl_core/pull/3437 (I am going to fix CI build failures by the way).
sdl_hmi PR: https://github.com/smartdevicelink/sdl_hmi/pull/436

**Comment by joeljfischer** [comment]
@shiniwat the Core PR does not implement both proposals that are implemented by this PR and therefore cannot be used to test this PR. Please create two separate PRs for the two proposals that you currently implement in this one PR.

**Comment by shiniwat** [comment]
Hi @joeljfischer, I have reverted the implementation commit for SDL 0323. Thanks.

**Comment by shiniwat** [comment]
Hi @NicoleYarroch, I thought all change requests are resolved a while back, but forgot to re-request the review. I made that right before, thanks.

**Review by NicoleYarroch — CHANGES_REQUESTED** [review]
I requested a few changes to the code to fix breaking changes.

**Comment by NicoleYarroch** [comment]
@shiniwat Please tag me when you are ready for Livio to re-review this PR. Thanks!

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h:49]: That makes sense. Thanks for pointing it out.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m
Comment (shiniwat) [reply SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m:82]: Understood. Updated SDLVideoStreamingCapabilitySpec.m accordingly.

File: SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m
Comment (shiniwat) [reply SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m]: Removed the extra white space.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h]: Understood that this comments needs to be created by rpc_spec generator.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h]: Updated the comment accordingly.

File: SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m
Comment (shiniwat) [reply SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m]: Removed the additional SDLLogD, and merged with existing one.

**Comment by NicoleYarroch** [comment]
@shiniwat I see that you made my requested changes. Just checking in to see if this PR is ready for me to review. Thanks!

**Comment by shiniwat** [comment]
Hi @NicoleYarroch,
Yes, I have made all changes you requested. I was wondering, however, why building unit test (SmartDeviceLinkTests) failed. While it's most unlikely relevant to my changes, I thought I should take a look prior to requesting the review.
On my dev box, I was running into another unit test error, which seems to be specific to XCode 12.3. Because I've been tied up with other tasks right now, I will take a look sometime later (in a day or two).

**Comment by NicoleYarroch** [comment]
@shiniwat Ah yes, there are known issues with building the tests on Xcode v.12.3. I have created a branch (https://github.com/smartdevicelink/sdl_ios/branches) with a workaround called "bugfix/Xcode_12.3_tests_not_building" if you would like to run the tests (please don't commit the changes on that branch since we expect an update to Carthage to fix this bug).

**Comment by shiniwat** [comment]
Hi @NicoleYarroch,
I have locally merged bugfix/Xcode_12.3_tests_not_building branch, and confirmed that solved the test builds failure. I believe this PR is ready for review, then. Thanks.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.h:49]: We are trying to use the inits generated by the RPC Spec generator script. That init allows us to pass a `nil` value for `preferredFPS`.

Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.h:49]:
```suggestion
/**
 * @param preferredResolution - preferredResolution
 * @param maxBitrate - maxBitrate
 * @param supportedFormats - supportedFormats
 * @param hapticSpatialDataSupported - hapticSpatialDataSupported
 * @param diagonalScreenSize - diagonalScreenSize
 * @param pixelPerInch - pixelPerInch
 * @param scale - scale
 * @param preferredFPS - preferredFPS
 * @return A SDLVideoStreamingCapability object
 */
- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(nullable NSNumber<SDLUInt> *)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticSpatialDataSupported:(nullable NSNumber<SDLBool> *)hapticSpatialDataSupported diagonalScreenSize:(nullable NSNumber<SDLFloat> *)diagonalScreenSize pixelPerInch:(nullable NSNumber<SDLFloat> *)pixelPerInch scale:(nullable NSNumber<SDLFloat> *)scale preferredFPS:(nullable NSNumber<SDLUInt> *)preferredFPS;
```

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.m:25]: When the generated init is used (see comment I left in the header file) you can set `nil` for the `preferredFPS` so you don't need the `DEFAULT_FRAME_RATE` anymore: 

```
self = [self initWithPreferredResolution:preferredResolution maxBitrate:@(maxBitrate) supportedFormats:supportedFormats hapticSpatialDataSupported:@(hapticDataSupported) diagonalScreenSize:@(diagonalScreenSize) pixelPerInch:@(pixelPerInch) scale:@(scale) preferredFPS:nil];
```

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.h:16]: Can you revert this back to the way you had it in the first review? I think it makes sense for the `SDLStreamingVideoLifecycleManager` to manage the logic for setting the default frame rate value if `preferredFPS` is `nil`.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.h:106]:
```suggestion
/**
 * The preferred frame rate per second of the head unit. The mobile application / app library may take other factors into account that constrain the frame rate lower than this value, but it should not perform streaming at a higher frame rate than this value.
 * {"num_min_value": 0, "num_max_value": 2147483647}
 *
 * @added in SmartDeviceLink 7.1.0
 */
```

Comment (NicoleYarroch) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h:106]: I just updated the docs with the generated documentation for the parameter

**Review by NicoleYarroch — CHANGES_REQUESTED** [review]
I left a suggestion for the `SDLVideoStreamingCapability`  inits so `nil` can be set for `preferredFPS`

**Comment by NicoleYarroch** [comment]
@shiniwat  Please tag me when you are ready for Livio to do a re-review. Thanks!

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h:49]: Understood. I have updated SDLVideoStreamingCapability.h based on the generated code (updated preferredFPS relevant code only).

Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h]: Updated SDLVideoStreamingCapability.h based on the generated code as you suggested.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.m]: I have updated SDLVideoStreamingCapability.m based on the generated code as per suggestion.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h]: Understood. Updated accordingly.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h]: Updated SDLVideoStreamingCapability.h based on the generated code. Thanks for the review.

**Comment by shiniwat** [comment]
Hi @NicoleYarroch, I have updated SDLVideoStreamingCapability (.h and .m) based on the generated code, and updated the callers (which include test code) accordingly. I believe it's ready for re-review. Thanks.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.h:33]: This `init` is not implemented in the **.m** file (I think the new init was accidentally copied twice?). One of the test cases is failing due to this.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLVideoStreamingCapability.m:25]:
```suggestion
- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale {
    return [self initWithPreferredResolution:preferredResolution maxBitrate:@(maxBitrate) supportedFormats:supportedFormats hapticSpatialDataSupported:@(hapticDataSupported) diagonalScreenSize:@(diagonalScreenSize) pixelPerInch:@(pixelPerInch) scale:@(scale) preferredFPS:nil];
}
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m:92]:
```suggestion
        expect(testStruct.preferredFPS).to(beNil());
```

**Review by NicoleYarroch — CHANGES_REQUESTED** [review]
I found an issue with one of the `SDLVideoStreamingCapability` inits that should be easily fixed. I also discovered something odd with comparing the `videoEncoderSettings` value with the `customEncoderSettings`, please let me know if you can reproduce it on your end since I'm not quite sure what is going on under the hood.

**Comment by NicoleYarroch** [comment]
@shiniwat Can you please tag me when this PR is ready for review? Thanks!

File: SmartDeviceLink/public/SDLVideoStreamingCapability.h
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.h:33]: Oops, I copied some `init` from generated SDLVideoStreamingCapability.m, but parameter set in generated method was different... Not sure why building the framework didn't fail on my XCode. My apologies.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m
Comment (shiniwat) [reply SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m]: Oops, thanks for catching this.

File: SmartDeviceLink/public/SDLVideoStreamingCapability.m
Comment (shiniwat) [reply SmartDeviceLink/public/SDLVideoStreamingCapability.m]: Fixed the init. Thanks.

File: SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m
Comment (shiniwat) [reply SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m]: Well, you're right that comparison should be explicitly made with NSNumber. Thanks for catching.

**Review by NicoleYarroch — APPROVED** [review]

**Pull request merged**
